### PR TITLE
scrubber: clean up `scan_metadata` before prod

### PR DESCRIPTION
**Pull request by yliang412**
Part of #8128.

## Problem
Currently, scrubber `scan_metadata` command will return with an error code if the metadata on remote storage is corrupted with fatal errors. To safely deploy this command in a cronjob, we want to differentiate between failures while running scrubber command and the erroneous metadata. At the same time, we also want our regression tests to catch corrupted metadata using the scrubber command.

## Summary of changes

- Return with error code only when the scrubber command fails
- Uses explicit checks on errors and warnings to determine metadata health in regression tests.

**Resolve conflict with `tenant-snapshot` command (after shard split):** [`test_scrubber_tenant_snapshot`](https://github.com/neondatabase/neon/blob/yuchen/scrubber-scan-cleanup-before-prod/test_runner/regress/test_storage_scrubber.py#L23) failed before applying  422a8443ddb7f1c7a26907a96c4aed0c5d554e67
- When taking a snapshot, the old `index_part.json` in the unsharded tenant directory is not kept. 
- The current `list_timeline_blobs` implementation consider no `index_part.json` as a parse error.
- During the scan, we are only analyzing shards with highest shard count, so we will not get a parse error. but we do need to add the layers to tenant object listing, otherwise we will get index is referencing a layer that is not in remote storage error.
- **Action:** Add s3_layers from `list_timeline_blobs` regardless of parsing error

_Open Question:_ What do we want to do with other commands that uses `list_timeline_blobs`?

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: storage_scrubber/src/checks.rs [hunk 1]
```diff
@@ -300,7 +303,10 @@ pub(crate) enum BlobDataParseResult {
     },
     /// The remains of a deleted Timeline (i.e. an initdb archive only)
     Relic,
-    Incorrect(Vec<String>),
+    Incorrect {
```
Comment (yliang412) [inline storage_scrubber/src/checks.rs:306]: We could also do an MissingIndex variant designated for this use.

File: storage_scrubber/src/scan_pageserver_metadata.rs [hunk 2]
```diff
@@ -290,13 +290,21 @@ pub async fn scan_metadata(
...
+                s3_layers,
+            } => {
+                tenant_objects.push(ttid, s3_layers.clone());
+            }
+            BlobDataParseResult::Relic => (),
+            BlobDataParseResult::Incorrect {
+                errors: _,
+                s3_layers,
+            } => {
+                tenant_objects.push(ttid, s3_layers.clone());
```
Comment (yliang412) [inline storage_scrubber/src/scan_pageserver_metadata.rs:306]: This is where we need add the layers in.

File: storage_scrubber/src/pageserver_physical_gc.rs [hunk 3]
```diff
@@ -389,10 +389,13 @@ async fn gc_ancestor(
                 // Post-deletion tenant location: don't try and GC it.
                 continue;
             }
-            BlobDataParseResult::Incorrect(reasons) => {
+            BlobDataParseResult::Incorrect {
+                errors,
+                s3_layers: _, // TODO(yuchen): could still check references to these s3 layers?
```
Comment (yliang412) [inline storage_scrubber/src/pageserver_physical_gc.rs:394]: I think for this case we could also check the ref count for these s3 layers?

File: storage_scrubber/src/main.rs [hunk 4]
```diff
@@ -208,21 +208,21 @@ async fn main() -> anyhow::Result<()> {
                         }
 
                         if summary.is_fatal() {
-                            Err(anyhow::anyhow!("Fatal scrub errors detected"))
+                            eprintln!("Fatal scrub errors detected");
```
Comment (skyzh) [inline storage_scrubber/src/main.rs:211]: should be an `error!` log line?

**Review by skyzh — APPROVED** [review]

File: storage_scrubber/src/main.rs (hunk 4 above)
Comment (yliang412) [reply storage_scrubber/src/main.rs]: 6b664176848781120da6a4951a7b183c3e73114d

**Review by arpad-m — APPROVED** [review]

File: storage_scrubber/src/pageserver_physical_gc.rs (hunk 3 above)
Comment (yliang412) [reply storage_scrubber/src/pageserver_physical_gc.rs:394]: Created https://github.com/neondatabase/neon/issues/8618

**Pull request merged**